keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
+  - otherwise: Deployment A is deleted immediately
+- Background cascading deletion: Deployment A is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to some of objects in a namespace, a deletion request to the namespace isn't blocked.
+This means that the resources with lien in the namespace will eventually be deleted by namespace lifecycle controller.
+Users can request to delete its namespace or other objects in the namespace.
+
+In addition, users may expect that the order of resource deletions inside a namespace is guaranteed even when requested through the namespace deletion.
+For alpha, resources with liens are deleted in a nondeterministic order on the namespace deletion.
+Handling order of deletion is a beta blocker and won't be included in the initial implementation.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, the API server must block any request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp`.
+
+#### Race of removing/adding liens
+
+A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.
+
+To mitigate the risk, a new `IgnoreLiens` API option in `DeleteOptions` to force delete a resource with liens will be added.
+
+Specifying the option should only be allowed to a limited set of users and groups, therefore deletion request will be blocked in api-server, if invalid users request deletion with the option.
+Restricting the `IgnoreLiens` API option to specific users is beta blocker and won't be included in the initial implementation.
```

Comment:
I think getting the permissions model right for alpha is important.  We don't want to needlessly break clients during an alpha to beta transition if we know this is coming.

Could this check actually be, "do you have the rights to modify the /liens subresource" since such permissions imply the ability to remove liens already and deleteoptions.ignoreLiens exists to eliminate the race.

Reply:
How about making DELETE permission for `{resource}/liens` subresource a requirement for `IgnoreLiens` API option (It may only be valid when DELETE permission isn't used for a normal use for subresourcing, but currently only GET, PATCH, and PUT verbs are planed to be used in [Subresource finalizers KEP](https://github.com/kubernetes/enhancements/issues/3285))?

Reply:
> How about making DELETE permission for {resource}/liens subresource a requirement for IgnoreLiens API option..?

I like the overall idea here; I want to check that we would restrict at the Kubernetes verb level not HTTP. For example, you might use Server-Side Apply to remove all Liens (HTTP verb `PATCH`); RBAC verb might be `update` (repeatedly removing one lien) or `delete`, I'm not sure.

Reply:
Is there a basic assumption that any user who can UPDATE/PATCH an object can clear liens?  If sho, we should say that.  Or do we need the ability to say "mere mortals cannot do that"?  E.g. some critical thing that must not be deleted until ACK'ed by a super-privileged actor?  I can't come up with any examples right now...

Assuming the above (any update/patch can clear liens), does the proposed check (delete `{resource}/liens`) pass for such users (e.g. I don't explicitly have permission to delete foo/liens but I do have permission to update foo) ?